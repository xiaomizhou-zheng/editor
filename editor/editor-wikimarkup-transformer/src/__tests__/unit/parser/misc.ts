import WikiMarkupTransformer from '../../../index';

describe('JIRA wiki markup - Misc', () => {
  const testCases: Array<[string, string]> = [
    [
      'should find emojis in text',
      'this is a string with :) emojis in it (*) tada',
    ],
    [
      'should find emojis in text with marks',
      'this is a string ~with :) emojis~ in it (*) tada',
    ],
    ['an emoji itself should work too', ':)'],
    [':puppy: should not be an emoji clusterfuck', ':puppy:'],
    ['an emoji with characters after should not be an emoji', ':)hello'],
    ['an emoji with characters before should still be an emoji', 'hello:)'],
    ['an emoji at the end of line should still be emoji', 'hello:)\n'],
    [
      'an emoji with characters before and after should not be an emoji',
      'hello:)hello',
    ],
    [
      'an emoji with characters before and after with a whitespace should be an emoji',
      'hello:) hello',
    ],
    [
      'should find emojis and mentions in text',
      'this is a string with :) emojis and [~username] mentions',
    ],
    ['should escape macros', 'this is a \\{panel} text, not a macro{panel}'],
    [
      'should escape monospace text',
      'this is a \\{{normal text}}, not a monospaced',
    ],
    [
      '[CS-480] this is not a list',
      `-
this is not a list`,
    ],
    [
      'It should be possible to have a ruler as the final line of text',
      `Hi
----`,
    ],
    ['It should be possible to have a ruler as the only thing', '----'],
    [
      'ESS-2440 should render consecutive emojis correctly',
      'emoji prefix (/)(*)(i) emoji suffix',
    ],
  ];

  for (const [testCaseDescription, markup] of testCases) {
    it(testCaseDescription, () => {
      const transformer = new WikiMarkupTransformer();
      expect(transformer.parse(markup)).toMatchSnapshot();
    });
  }
});
