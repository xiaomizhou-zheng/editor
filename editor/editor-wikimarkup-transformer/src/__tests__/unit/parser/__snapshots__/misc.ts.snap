// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JIRA wiki markup - Misc :puppy: should not be an emoji clusterfuck 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": ":puppy:",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Misc [CS-480] this is not a list 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "-",
          "type": "text",
        },
        Object {
          "type": "hardBreak",
        },
        Object {
          "text": "this is not a list",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Misc ESS-2440 should render consecutive emojis correctly 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "emoji prefix ",
          "type": "text",
        },
        Object {
          "attrs": Object {
            "id": "atlassian-check_mark",
            "shortName": ":check_mark:",
            "text": ":check_mark:",
          },
          "type": "emoji",
        },
        Object {
          "attrs": Object {
            "id": "atlassian-yellow_star",
            "shortName": ":yellow_star:",
            "text": ":yellow_star:",
          },
          "type": "emoji",
        },
        Object {
          "attrs": Object {
            "id": "atlassian-info",
            "shortName": ":info:",
            "text": ":info",
          },
          "type": "emoji",
        },
        Object {
          "text": " emoji suffix",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Misc It should be possible to have a ruler as the final line of text 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "Hi",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "type": "rule",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Misc It should be possible to have a ruler as the only thing 1`] = `
Object {
  "content": Array [
    Object {
      "type": "rule",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Misc an emoji at the end of line should still be emoji 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "hello",
          "type": "text",
        },
        Object {
          "attrs": Object {
            "id": "1f642",
            "shortName": ":slight_smile:",
            "text": "ðŸ™‚",
          },
          "type": "emoji",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Misc an emoji itself should work too 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "id": "1f642",
            "shortName": ":slight_smile:",
            "text": "ðŸ™‚",
          },
          "type": "emoji",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Misc an emoji with characters after should not be an emoji 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": ":)hello",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Misc an emoji with characters before and after should not be an emoji 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "hello:)hello",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Misc an emoji with characters before and after with a whitespace should be an emoji 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "hello",
          "type": "text",
        },
        Object {
          "attrs": Object {
            "id": "1f642",
            "shortName": ":slight_smile:",
            "text": "ðŸ™‚",
          },
          "type": "emoji",
        },
        Object {
          "text": " hello",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Misc an emoji with characters before should still be an emoji 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "hello",
          "type": "text",
        },
        Object {
          "attrs": Object {
            "id": "1f642",
            "shortName": ":slight_smile:",
            "text": "ðŸ™‚",
          },
          "type": "emoji",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Misc should escape macros 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "this is a {panel} text, not a macro",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "attrs": Object {
        "panelColor": null,
        "panelIcon": null,
        "panelIconId": null,
        "panelIconText": null,
        "panelType": "info",
      },
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "panel",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Misc should escape monospace text 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "this is a {{normal text}}, not a monospaced",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Misc should find emojis and mentions in text 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "this is a string with ",
          "type": "text",
        },
        Object {
          "attrs": Object {
            "id": "1f642",
            "shortName": ":slight_smile:",
            "text": "ðŸ™‚",
          },
          "type": "emoji",
        },
        Object {
          "text": " emojis and ",
          "type": "text",
        },
        Object {
          "attrs": Object {
            "accessLevel": "",
            "id": "username",
            "text": "",
            "userType": null,
          },
          "type": "mention",
        },
        Object {
          "text": " mentions",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Misc should find emojis in text 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "this is a string with ",
          "type": "text",
        },
        Object {
          "attrs": Object {
            "id": "1f642",
            "shortName": ":slight_smile:",
            "text": "ðŸ™‚",
          },
          "type": "emoji",
        },
        Object {
          "text": " emojis in it ",
          "type": "text",
        },
        Object {
          "attrs": Object {
            "id": "atlassian-yellow_star",
            "shortName": ":yellow_star:",
            "text": ":yellow_star:",
          },
          "type": "emoji",
        },
        Object {
          "text": " tada",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Misc should find emojis in text with marks 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "this is a string ",
          "type": "text",
        },
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "type": "sub",
              },
              "type": "subsup",
            },
          ],
          "text": "with ",
          "type": "text",
        },
        Object {
          "attrs": Object {
            "id": "1f642",
            "shortName": ":slight_smile:",
            "text": "ðŸ™‚",
          },
          "type": "emoji",
        },
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "type": "sub",
              },
              "type": "subsup",
            },
          ],
          "text": " emojis",
          "type": "text",
        },
        Object {
          "text": " in it ",
          "type": "text",
        },
        Object {
          "attrs": Object {
            "id": "atlassian-yellow_star",
            "shortName": ":yellow_star:",
            "text": ":yellow_star:",
          },
          "type": "emoji",
        },
        Object {
          "text": " tada",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;
